src/shell/js/composeui-fdc3/src/ComposeUIDesktopAgent.ts

```diff
@@ -0,0 +1,312 @@
+    }
+
+    //TODO
+    public addIntentListener(intent: string, handler: IntentHandler): Promise<Listener> {
+        throw new Error("Not implemented");
+    }
+
+    public addContextListener(contextType?: string | null | ContextHandler, handler?: ContextHandler): Promise<Listener> {
+        return new Promise<ComposeUIListener>(async (resolve, reject) => {
+            if (!this.currentChannel) {
+                reject(new Error("The current channel have not been set."));
+                return;
+            }
+            if (typeof contextType != 'string' || !contextType) {
+                reject(new Error("The contextType was type of ContextHandler, which would use a deprecated function, please use string or null for contextType!"));
+                return;
+            }
+
+            const stringContextType = contextType as string ?? null;
+
+            const listener = <ComposeUIListener>await this.currentChannel!.addContextListener(stringContextType, handler!);
+            await this.currentChannel!.getCurrentContext(stringContextType)
+                .then(async (resultContext) => {
+                    listener.LatestContext = this.currentChannel!.retrieveCurrentContext(stringContextType);
+                    if (resultContext != listener.LatestContext) {
+                        //TODO: integrationtest
+                        await listener.handleContextMessage();
+                    } else {
+                        await listener.handleContextMessage(resultContext);
+                    }
+                });
+
+            this.currentChannelListeners.push(listener);
+            resolve(listener);
+        });
+    }
+
+    public getUserChannels(): Promise<Array<Channel>> {
+        return Promise.resolve(this.userChannels);
+    }
+
+    //TODO: should return AccessDenied error when a channel object is denied?
+    public joinUserChannel(channelId: string): Promise<void> {
+        return new Promise<void>(async (resolve, reject) => {
+            if (this.currentChannel) {
+                reject(new Error(ChannelError.AccessDenied));
+            }
+
+            let channel = this.userChannels.find(innerChannel => innerChannel.id == channelId);
+            if (!channel) {
+                try{
+                    await this.invokeChannelCreationMessage(ComposeUITopic.joinUserChannel(), channelId, "user");
+                    resolve();
+                } catch(error) {
+                    reject(error);
+                }
+            } else {
+                this.currentChannel = channel;
+                resolve();
+            }
+        });
+    }
+
+    //TODO: should return AccessDenied error when a channel object is denied
+    //TODO: should return a CreationFailed error when a channel cannot be created or retrieved (channelId failure)
+    public getOrCreateChannel(channelId: string): Promise<Channel> {
+        // return new Promise<Channel>(async(resolve, reject) => {
```

Comment:
Is this not supported by the backend? Why is it commented out?

Reply:
We agreed on that if that is not supported by the backend then I should comment out / remove the code. I decided to comment out.